### DISCUSSION TITLE
Make it easier to have a pocketbase base directory

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#3958] [Q&A]
Currently the `--dir` is hard coded into `pocketbase.go` as the data directory.  This makes it fairly awkward to set a base directory for pocketbase.  Now there are migrations, hooks, and public.  It would be nice if pocketbase.go had built into it the concept of a base directory.

As you have it if there was a `--base-dir` that was set, then inside of it would appear `data` directory.  Then someone could add `migrations`, `hooks`, and `public`.

So if `--base-dir` was not set then pocketbase would function is it is now.  Otherwise it would be a relative path within baseDir.

This would allow me to piggy-back off of the baseDir for the other features that I'm adding in my PB as Framework custom build.  For the moment that is just the directories that your already using in your default binary.  Base directory could be used for custom `ssh`, some `keys`, etc.

------

I also think that `--dir` is not a great name for the DB directory (now that PB has three other relevant directories), but that it's probably not worth renaming it to `--data-dir` for how much that it would fail everyone.

**Comment by dschissler** [comment]
I understand that this handles an edge case where someone created the hooks after pocketbase is running, but it seems like that it should be stored in the hooks directory by default.

```javascript
	// If not set it fallbacks to "pb_data".
	TypesDir string
```

**Reply by dschissler** [reply]
Oops, I see why you put it in data directory.  The JSVM watcher restarts  the app in a loop after it detects the writing of the `types.d.ts` file.  

You should probably at least change the ordering of it so that the types file is written before the watcher starts.  This seems like a bug fix that could be released before `0.21.0`.

**Reply by ganigeorgiev** [reply]
What the files watcher has to do with the ambient TS declarations?

**Reply by dschissler** [reply]
The JSVM (or something) is writing `types.d.ts` on startup.  If that directory is the hooks directory, and if hooks watching is on then it triggers an application `Restart`, which puts it into an infinite loop.

**Reply by dschissler** [reply]
`File /home/pi/project/hooks/types.d.ts changed, restarting...`

**Reply by dschissler** [reply]
My thinking is that since you're only providing `__hooks` directory constant that everything should probably just be based out of the hooks directory.  I think of `__hooks` more of like `__jsvm` in that it's the place where that entire portion of the app takes place.

**Reply by ganigeorgiev** [reply]
I understand. I've added in my todo to consider updating the firing order with the next release but even as it is now it shouldn't be a blocker or critical issue. The types file is in `pb_data` because it is a shared/common directory that both `pb_migrations` and `pb_hooks` have access to and also `pb_data` is expected to be gitignored, otherwise the types could get updated with the binary start and will trigger an unnecessary git change.

**Reply by dschissler** [reply]
I noticed that new `.pb.js` files are having `/// <reference path="../types.d.ts" />` automatically added to them.  I assume that it would resolve them deeper through some algorithm.

I think that PB should be done with one of two styles (that everyone can choose)

1) `data` in its own directory, and everything else in a base directory, so `base_dir_name/hooks`, `base_dir_name/migrations`, `base_dir_name/public`, etc.

2) `base_dir_name`/ everything goes in here.  This whole directory can be a git repo that has `data` in gitignore.

Probably the first one would be best.

Or maybe the approach that your suggesting is that the user changes directory into a base directory and then everything just happens from there automatically.  I suppose that since I'm always building it as a framework that I just never thought of that.  This would be more intuitive to someone who was using the pre-built binary.

In that case everything is fine.  Maybe everything would be just easier if I just changed directories like that.

Also though, you could offer a `--base-dir` option that would be essentially like changing into a directory before running pocketbase.  If you don't want to do that then perhaps the Documentation could be more explicit about how its best to run Pocketbase.

I'm probably stretching Pocketbase more than 99% of the users in this way since I'm using it as a Golang and Javascript framework at the same time.  Things got slightly weird.

**Reply by dschissler** [reply]
I imagined that I was being silly and overcomplicated with my build too.  I tried to set the directory that a build task runs in, but then I ran into `go run` not being able to find my code.

I attempted to just resign to the simple life, but it's not actually so simple.  I think that we could use a `--base-dir`.  Then if `--dir` was set as a relative path then it would be named that inside `--base-dir`, otherwise it gets its absolute path.

I suppose that in the default app that all of these other directories can be made to function the same if a base directory is set.
